### PR TITLE
[BACK-1746] Convert string dates to ISODATE in DB

### PR DESCRIPTION
**Pull request by Roukoswarf**
This has been tested using uploader and postman APITests along with manual testing, and covers all found cases using string dates.
Unit tests will most likely fail for now, as they manually specify date strings, and I believe it will be best to gather feedback on the approach, and discuss unit testing strategies before adjusting them.

### CODE REVIEW

File: store/store.go [hunk 1]
```diff
@@ -484,7 +534,7 @@ func (c *MongoStoreClient) GetDeviceData(p *Params) (StorageIterator, error) {
 
 	// _schemaVersion is still in the list of fields to remove. Although we don't query for it, data can still exist for it
 	// until BACK-1281 is done.
-	removeFieldsForReturn := bson.M{"_id": 0, "_userId": 0, "_groupId": 0, "_version": 0, "_active": 0, "_schemaVersion": 0, "createdTime": 0, "modifiedTime": 0}
+	removeFieldsForReturn := bson.M{"_id": 0, "_userId": 0, "_groupId": 0, "_version": 0, "_active": 0, "_schemaVersion": 0, "createdTime": 0, "modifiedTime": 0, "_deduplicator": 0}
```
Comment (toddkazakov) [inline store/store.go:537]: Is this change intended?

File: tide-whisperer.go [hunk 2]
```diff
@@ -78,8 +80,9 @@ var (
 
 const (
 	dataAPIPrefix             = "api/data "
-	medtronicLoopBoundaryDate = "2017-09-01"
+	medtronicLoopBoundaryDateUnix = 1504238400
```
Comment (toddkazakov) [inline tide-whisperer.go:83]: It doesn't make much sense to have a timestamp here, that's later converted to `time.Time`. Having a string is better for readability.

File: tide-whisperer.go [hunk 3]
```diff
@@ -330,6 +335,14 @@ func main() {
 			}
 
 			if len(results) > 0 {
+				// HACK convert deviceTime to string before marshal, to avoid modifying the results map above
+				switch v := results["deviceTime"].(type) {
```
Comment (toddkazakov) [inline tide-whisperer.go:339]: This doesn't need to be a `switch` statement, simple `if` should suffice.

**Review by toddkazakov — CHANGES_REQUESTED** [review]
I think this can be further simplified. I don't see a reason to convert query params from strings to time, only to convert them back to strings when querying mongo. I think you should keep the `MedronicDate` query param as a string in the params struct and convert to `Time` right before fetching the data from the db.

File: store/store.go (hunk 1 above)
Comment (Roukoswarf) [reply store/store.go:537]: yes, when testing the clients, tide-whisperer was leaving extra properties in the return, likely just forgotten, seemed like some minor cleanup.
It can be reverted, if needed. Its an internal property that shouldnt have been hitting clients from what i can tell.

File: store/store.go [hunk 4]
```diff
@@ -80,31 +82,35 @@ type (
...
+		return date, nil
 	}
+
 	date, err := time.Parse(time.RFC3339Nano, dateString)
 	if err != nil {
-		return "", err
+		return date, err
 	}
 
 	// The Golang implementation of time.RFC3339Nano does not use a fixed number of digits after the
```
Comment (toddkazakov) [inline store/store.go:97]: This comment should be moved to the appropriate place

File: store/store.go [hunk 5]
```diff
@@ -208,6 +214,7 @@ func (c *MongoStoreClient) WithContext(ctx context.Context) *MongoStoreClient {
 // EnsureIndexes exist for the MongoDB collection. EnsureIndexes uses the Background() context, in order
 // to pass back the MongoDB errors, rather than any context errors.
 func (c *MongoStoreClient) EnsureIndexes() error {
+	//medtronicIndexDateTime, _ := time.Parse(medtronicDateFormat, medtronicIndexDate)
```
Comment (toddkazakov) [inline store/store.go:217]: Should this be removed?

File: store/store.go (hunk 1 above)
Comment (toddkazakov) [reply store/store.go:543]: Although it's internal, there might be 3rd party clients which depend on it.

**Review by toddkazakov — CHANGES_REQUESTED** [review]
Looks good overall. There are some small issues and tests must be fixed.

File: store/store.go [hunk 6]
```diff
@@ -222,7 +229,7 @@ func (c *MongoStoreClient) EnsureIndexes() error {
 							"$exists": true,
 						}},
 						{Key: "time", Value: bson.M{
-							"$gte": "2017-09-01",
+							"$gte": medtronicIndexDate,
```
Comment (toddkazakov) [inline store/store.go:232]: Do we need a new index that covers native dates?

File: store/store.go [hunk 7]
```diff
@@ -237,7 +244,7 @@ func (c *MongoStoreClient) EnsureIndexes() error {
 						{Key: "_active", Value: true},
 						{Key: "origin.payload.device.manufacturer", Value: "Medtronic"},
 						{Key: "time", Value: bson.M{
-							"$gte": "2017-09-01",
+							"$gte": medtronicIndexDate,
```
Comment (toddkazakov) [inline store/store.go:247]: Do we need a new index that covers native dates?

File: store/store.go [hunk 8]
```diff
@@ -446,10 +536,21 @@ func (c *MongoStoreClient) GetLoopableMedtronicDirectUploadIdsAfter(userID strin
...
+	}
+	if err != nil {
+		return nil, errors.New("date is invalid")
+	}
+
 	query := bson.M{
 		"_active":     true,
 		"_userId":     userID,
-		"time":        bson.M{"$gte": date},
+		"$or": bson.A{
```
Comment (toddkazakov) [inline store/store.go:550]: Formatting

File: store/store.go [hunk 9]
```diff
@@ -80,31 +82,31 @@ type (
...
-	// See https://github.com/golang/go/issues/19635
-	return date.Format(RFC3339NanoSortable), nil
+	return date, nil
 }
 
 // GetParams parses a URL to set parameters
 func GetParams(q url.Values, schema *SchemaVersion) (*Params, error) {
 
-	startStr, err := cleanDateString(q.Get("startDate"))
+	// NOTE: *Str is redundant once dates are all converted to native dates
```
Comment (toddkazakov) [inline store/store.go:103]: Should this comment be removed?

File: store/store.go [hunk 10]
```diff
@@ -237,7 +241,41 @@ func (c *MongoStoreClient) EnsureIndexes() error {
...
 						{Key: "origin.payload.device.manufacturer", Value: "Medtronic"},
 						{Key: "time", Value: bson.M{
-							"$gte": "2017-09-01",
+							"$gte": medtronicIndexDate,
+						}},
+					},
+				),
+		},
+		// END LEGACY INDEXES
+		{
```
[the remark this answers is not in the document]
Comment (Roukoswarf) [reply store/store.go:250]: that explains a lot.

File: store/store.go (hunk 9 above)
Comment (Roukoswarf) [reply store/store.go:103]: possibly, i was expecting to note all double date handling so it could be cleaned up easier at a later date once no longer needed, but the number of locations made it mostly white noise to keep doing, ill remove them.

**Review by toddkazakov — APPROVED** [review]
Don't forget to create the indexes in production before deploying this
